[Docs] Translate infer.md

Hongru-Xiao opened the pull request:
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/open-mmlab/mmengine?pullRequest=1121) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/open-mmlab/mmengine?pullRequest=1121) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/open-mmlab/mmengine?pullRequest=1121) it.</sub>

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
````diff
@@ -0,0 +1,244 @@
+    ...
+```
+
+For specific details, please refer to the [Development Specification](#Development Specification of Inference Interface) below.
````
Please make sure the link is available

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+
+For specific details, please refer to the [Development Specification](#Development Specification of Inference Interface) below.
+
+### Building an Inference Engine
```
```suggestion
### Building an Inferencer
```

HAOCHENYE replied:
Please replace all `inference engine` with `Inferencer` in this document

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+
+### Building an Inference Engine
+
+**Building an Inference Engine Based on Configuration File Path**
```
```suggestion
**Building an Inference Engine Based on Configuration File**
```

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
````diff
@@ -0,0 +1,244 @@
+inferencer = CustomInferencer(model=cfg, weight=weight)
+```
+
+**Building an Inference Engine Based on Model Name Defined in 'model - index**，Taking the [ATSS detector in MMDetection](https://github.com/open-mmlab/mmdetection/blob/31c84958f54287a8be2b99cbf87a6dcf12e57753/configs/atss/metafile.yml#L22) as an example , the model name is `atss_r50_fpn_1x_coco`, since the weight path is already defined in the model-index, the weight parameter does not need to be ...
````
```suggestion
**Building an Inferencer based on model name defined in 'model - index**. Take the [ATSS detector in MMDetection](https://github.com/open-mmlab/mmdetection/blob/31c84958f54287a8be2b99cbf87a6dcf12e57753/configs/atss/metafile.yml#L22) as an example, the model name is `atss_r50_fpn_1x_coco`, since the weight path is already defined in the model-index, the weight parameter does not need to be configured.
```

Please refine

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+
+### Performing Inference
+
+**Inferring on a Single Image**
```
```suggestion
**Infer a single image**
```

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+
+## Development Specification of Inference Interface
+
+When inferencer performing inference, the following steps are typically executed：
```
```suggestion
When performing inference, the following steps are typically executed：
```

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+
+When inferencer performing inference, the following steps are typically executed：
+
+1. preprocess：Input data preprocessing, including data reading, data preprocessing, data format conversion, and so on, is typically carried out when performing inference.
+2. forward: Model Forward Inference.
+3. visualize：Visualization of Predicted Results.
+4. postprocess：Post-processing of predicted results, including result format conversion, exporting predicted results, and so on.
```
```suggestion
1. preprocess：Input data preprocessing, including data reading, data preprocessing, data format conversion, and so on.
2. forward: Execute model`.forwward`
3. visualize：Visualization of predicted results.
4. postprocess：Post-processing of predicted results, including result format conversion, exporting predicted results, and so on.
```

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+3. visualize：Visualization of Predicted Results.
+4. postprocess：Post-processing of predicted results, including result format conversion, exporting predicted results, and so on.
+
+To optimize the user experience of the inferencer, we aim to eliminate the need for users to configure parameters for each step when performing inference. In other words, we want users to be able to simply configure parameters for the `__call__` interface without having to be aware of the aforementioned processes, in order to complete inference.
```
```suggestion
To optimize the user experience of the inferencer, we aim to eliminate the need for users to configure parameters for each step when performing inference. In other words, we want users to be able to simply configure parameters for the `__call__` interface without having to be aware of the aforementioned processes.
```

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+
+To optimize the user experience of the inferencer, we aim to eliminate the need for users to configure parameters for each step when performing inference. In other words, we want users to be able to simply configure parameters for the `__call__` interface without having to be aware of the aforementioned processes, in order to complete inference.
+
+The `__call__` interface will execute the aforementioned steps in order, but it is not aware of which step the parameters provided by the user should be distributed to. Therefore, when developing a `CustomInferencer`, developers need to define four class attributes: `preprocess_kwargs`, `forward_kwargs`, `visualize_kwargs`, and `postprocess_kwargs`. Each attribute is a set of strings (`Set[str]`) ...
```
```suggestion
The `__call__` interface will execute the aforementioned steps in order, but it is not aware of which step the parameters provided by the user should be distributed to. Therefore, when developing a `CustomInferencer`, developers need to define four class attributes: `preprocess_kwargs`, `forward_kwargs`, `visualize_kwargs`, and `postprocess_kwargs`. Each attribute is a set of strings that are used to specify which step the parameters in the `__call__` interface correspond to:
```

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+
+### `__init__()`
+
+The `BaseInferencer.__init__` method has already implemented the logic for building an inferencer as shown in the [sample code](#Building an Inference Engine) , so in most cases there is no need to override the `__init__` method. However, if there is a need to implement custom logic for loading configuration files, weight initialization, pipeline initialization, etc., the `__init__` method can be ...
```
Check the link

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+
+### `_init_visualizer()`
+
+Initializes and returns the `visualizer` required by the inferencer, which is equivalent to the `visualizer` used in the training process. By default, `BaseInferencer` obtains the `visualizer` from the configuration of the `visualizer`, so there is usually no need to rewrite the `_init_visualizer` function.
```
```suggestion
Initializes and returns the `visualizer` required by the inferencer, which is equivalent to the `visualizer` used in the training process. By default, `BaseInferencer` obtains the `visualizer` from the configuration of the `visualizer`, so there is usually no need to overide the `_init_visualizer` function.
```

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+
+Output: 
+
+- Generator that yields one batch of data per iteration.。
```
```suggestion
- Generator that yields one batch of data per iteration.
```

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+
+### `preprocess()`
+
+Input parameters: 
```
Replace Input parameters with Input arguements

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+- inputs：The input data, which is the raw data without preprocessing.
+- preds：Predicted results of the model.
+- show：Whether to visualize.
+- Other parameters: specified by the user in `visualize_kwargs`.
```
```suggestion
- Other parameters: Specified by the user in `visualize_kwargs`.
```

Hongru-Xiao commented:
Hello, is there any other modification I need to make?

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+# Inference interface
+
+When developing based on MMEngine, we usually define a configuration file for a specific algorithm, use the configuration file to build an [executor](./runner.md), and execute the training and testing processes, and save the trained weights. When performing inference based on the trained model, the following steps are usually required:
```
```suggestion
When developing based on MMEngine, we usually define a configuration file for a specific algorithm, use the configuration file to build an [runner](./runner.md), and execute the training and testing processes, and save the trained weights. When performing inference based on the trained model, the following steps are usually required:
```

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
````diff
@@ -0,0 +1,244 @@
+inferencer = CustomInferencer(model=cfg, weight=weight)
+```
+
+**Building an Inferencer based on model name defined in 'model - index**. Take the [ATSS detector in MMDetection](https://github.com/open-mmlab/mmdetection/blob/31c84958f54287a8be2b99cbf87a6dcf12e57753/configs/atss/metafile.yml#L22) as an example, the model name is `atss_r50_fpn_1x_coco`, since the weight path is already defined in the model-index, the weight parameter does not need to be configu ...
````
```suggestion
**Building an Inferencer based on model name defined in model-index**. Take the [ATSS detector in MMDetection](https://github.com/open-mmlab/mmdetection/blob/31c84958f54287a8be2b99cbf87a6dcf12e57753/configs/atss/metafile.yml#L22) as an example, the model name is `atss_r50_fpn_1x_coco`. Since the path of weight has already been defined in the model-index, we don't need to configure the weight arguments anymore.
```

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+
+### Performing Inference
+
+**Infer a Single Image**
```
```suggestion
**Inferring on a single image**
```

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+When performing inference, the following steps are typically executed：
+
+1. preprocess：Input data preprocessing, including data reading, data preprocessing, data format conversion, and so on.
+2. forward: Execute model`.forwward`
```
```suggestion
2. forward: Execute `model.forwward`
```

HAOCHENYE commented on docs/zh_cn/design/infer_translation.md:
````diff
@@ -0,0 +1,244 @@
+In OpenMMLab's algorithm repositories, the Inferencer must be registered to the downstream repository's registry instead of the root registry of MMEngine to avoid naming conflicts.
+```
+
+**Core Interface Explanation**:
````
It could be better to make it as a level 2 heading. Please also refine the Chinese doc

Hongru-Xiao replied:
Excuse me, is the English translation also named 'infer.md'? Will it conflict with Chinese doc?

HAOCHENYE reviewed (commented):
Hi, once you have addressed the comments, please feel free to resolve them.

HAOCHENYE commented:
BTW, please follow the guide in [contributing.md](https://github.com/open-mmlab/mmengine/blob/main/CONTRIBUTING.md) to fix the lint

OpenMMLab-Assistant-004 commented:
Hi @Hongru-Xiao,

We'd like to express our appreciation for your valuable contributions to the mmengine. Your efforts have significantly aided in enhancing the project's quality.
It is our pleasure to invite you to join our community thorugh Discord_Special Interest Group (SIG) channel. This is a great place to share your experiences, discuss ideas, and connect with other like-minded people. To become a part of the SIG channel, send a message to the moderator, OpenMMLab, briefly introduce yourself and mention your open-source contributions in the #introductions channel. Our team will gladly facilitate your entry. We eagerly await your presence. Please follow this link to join us: ​https://discord.gg/UjgXkPWNqA.

If you're on WeChat, we'd also love for you to join our community there. Just add our assistant using the WeChat ID: openmmlabwx. When sending the friend request, remember to include the remark "mmsig + Github ID".

Thanks again for your awesome contribution, and we're excited to have you as part of our community!

Hongru-Xiao commented:
I have made the requested modifications and would appreciate it if you could review them.

HAOCHENYE commented:
> Hi, once you have addressed the comments, please feel free to resolve them.

Hi, once you have addressed the comments, please feel free to resolve them :smile: .

Hongru-Xiao commented:
> > Hi, once you have addressed the comments, please feel free to resolve them.
> 
> Hi, once you have addressed the comments, please feel free to resolve the conversation 😄 .
> 
> ![image](https://user-images.githubusercontent.com/57566630/236751774-c56db620-7314-414a-a5bb-10525d4c2ddb.png)

Oops, thank you for pointing that out. I have resolved all conversations.

HAOCHENYE commented:
@Hongru-Xiao 
Hi! Then lint seems still failed, Please follow the guide in [contributing.md](https://github.com/open-mmlab/mmengine/blob/main/CONTRIBUTING.md) to fix the lint error

zhouzaida commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+- inputs：The batch data processed by `preprocess` function.
+- Other parameters: Specified by the user and indicated in `forward_kwargs`.
+
+Output：
```
```suggestion
Return:
```

zhouzaida commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+Input arguments:
+
+- inputs：The batch data processed by `preprocess` function.
+- Other parameters: Specified by the user and indicated in `forward_kwargs`.
```
```suggestion
- Other parameters: Passed in by the user and specified in `forward_kwargs`.
```

zhouzaida commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+
+- inputs：Input data, passed into `__call__`, usually a list of image paths or image data.
+- batch_size：batch size, passed in by the user when calling `__call__`.
+- Other parameters: specified by the user and included in `preprocess_kwargs`.
```
```suggestion
- Other parameters: Passed in by the user and specified in `preprocess_kwargs`.
```

zhouzaida commented on docs/zh_cn/design/infer_translation.md:
````diff
@@ -0,0 +1,244 @@
+### `visualize()`
+
+```{note}
+Abstract method that subclasses must implement.
````
```suggestion
This is an abstract method that must be implemented by the subclass.
```

zhouzaida commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+- inputs：The input data, which is the raw data without preprocessing.
+- preds：Predicted results of the model.
+- show：Whether to visualize.
+- Other parameters: Specified by the user in `visualize_kwargs`.
```
```suggestion
- Other parameters: Passed in by the user and specified in `visualize_kwargs`.
```

zhouzaida commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+- show：Whether to visualize.
+- Other parameters: Specified by the user in `visualize_kwargs`.
+
+Output:
```
```suggestion
Return:
```

zhouzaida commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+
+Output:
+
+- The visualization result, usually of type `List[np.ndarray]`. Taking object detection as an example, each element in the list should be an image with detection boxes drawn, which can be visualized using `cv2.imshow`. The visualization process may vary for different tasks, and `visualize` should return results that are suitable for common visualization processes in that field.
```
```suggestion
- Visualize the results, which are usually of type `List[np.ndarray]`. Taking object detection as an example, each element in the list should be an image with detection boxes drawn, which can be visualized using `cv2.imshow`. The visualization process may vary for different tasks, and `visualize` should return results that are suitable for common visualization processes in that field.
```

zhouzaida commented on docs/zh_cn/design/infer_translation.md:
````diff
@@ -0,0 +1,244 @@
+### `postprocess()`
+
+```{note}
+Abstract method that subclasses must implement.
````
```suggestion
This is an abstract method that must be implemented by the subclass.
```

zhouzaida commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+- return_datasample：Whether to return the datasample as is. When set to `False`, the returned result is converted to a `dict`.
+- Other parameters: specified by the user in `postprocess_kwargs`.
+
+Output：
```
```suggestion
Return：
```

zhouzaida commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+- preds：The predicted results of the model, which is a `list` type. Each element in the list represents the prediction result of a piece of data. In the OpenMMLab algorithm library series, the type of each element in the prediction result is `BaseDataElement`.
+- visualization：Visualization results
+- return_datasample：Whether to return the datasample as is. When set to `False`, the returned result is converted to a `dict`.
+- Other parameters: specified by the user in `postprocess_kwargs`.
```
```suggestion
- Other parameters: Passed in by the user and specified in `postprocess_kwargs`.
```

zhouzaida commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+
+Input arguments:
+
+- preds：The predicted results of the model, which is a `list` type. Each element in the list represents the prediction result of a piece of data. In the OpenMMLab algorithm library series, the type of each element in the prediction result is `BaseDataElement`.
```
```suggestion
- preds：The predicted results of the model, which is a `list` type. Each element in the list represents the prediction result for a single data item. In the OpenMMLab series of algorithm libraries, the type of each element in the prediction result is `BaseDataElement`.
```

zhouzaida commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+- batch_size：Batch size for inference, which will be further passed to the `preprocess` function.
+- Extra arguments to be passed to `preprocess`, `forward`, `visualize`, and `postprocess` functions.
+
+Output:
```
```suggestion
Return:
```

zhouzaida commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+
+- preds：The predicted results of the model, which is a `list` type. Each element in the list represents the prediction result of a piece of data. In the OpenMMLab algorithm library series, the type of each element in the prediction result is `BaseDataElement`.
+- visualization：Visualization results
+- return_datasample：Whether to return the datasample as is. When set to `False`, the returned result is converted to a `dict`.
```
```suggestion
- return_datasample: Whether to maintain datasample for return. When set to `False`, the returned result is converted to a `dict`.
```

zhouzaida commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+Input arguments:
+
+- inputs：The input data, usually a list of image paths or image data. Each element in `inputs` can also be other types of data as long as it can be processed by the `pipeline` returned by [init_pipeline](#_init_pipeline). When there is only one inference data in `inputs`, it does not have to be a `list`, `__call__` will internally wrap it into a list for further processing.
+- return_datasample：Whether to return datasample as a `dict`.
```
```suggestion
- return_datasample: Whether to convert datasample to dict for return.
```

zhouzaida commented on docs/zh_cn/design/infer_translation.md:
```diff
@@ -0,0 +1,244 @@
+- inputs：The input data, usually a list of image paths or image data. Each element in `inputs` can also be other types of data as long as it can be processed by the `pipeline` returned by [init_pipeline](#_init_pipeline). When there is only one inference data in `inputs`, it does not have to be a `list`, `__call__` will internally wrap it into a list for further processing.
+- return_datasample：Whether to return datasample as a `dict`.
+- batch_size：Batch size for inference, which will be further passed to the `preprocess` function.
+- Extra arguments to be passed to `preprocess`, `forward`, `visualize`, and `postprocess` functions.
```
```suggestion
- Other parameters: Additional parameters assigned to `preprocess`, `forward`, `visualize`, and `postprocess` methods.
```

Hongru-Xiao commented:
Thanks again for the guidances of the reviewers. I have learned a lot from this PR and look forward to contributing to the project again in the future.
